source/common/router/retry_state_impl.cc

```diff
@@ -167,6 +167,16 @@ RetryStatus RetryStateImpl::shouldRetry(const Http::HeaderMap* response_headers,
 
 bool RetryStateImpl::wouldRetry(const Http::HeaderMap* response_headers,
                                 const Optional<Http::StreamResetReason>& reset_reason) {
+  // We never retry if the overloaded header is set.
+  if (response_headers && response_headers->EnvoyOverloaded() != nullptr) {
```

Comment:
Tiny nit: the first clause is a pointer and relies on implicit null checking, the second is also presumably a pointer and does an explicit `nullptr` comparison. So, might be nice to be consistent there.